banyan-cli/src/main.rs

```diff
@@ -30,142 +30,159 @@ pub type Result<T> = anyhow::Result<T>;
-            .required(true)
-            .takes_value(true)
-            .help("The root hash to use")
-    };
-    let topic_arg = || {
-        Arg::with_name("topic")
-            .long("topic")
-            .required(true)
-            .takes_value(true)
-            .help("The topic to send/recv data over")
-    };
-    let index_pass_arg = || {
-        Arg::with_name("index_pass")
-            .long("index_pass")
-            .required(false)
-            .takes_value(true)
-            .help("An index password to use")
-    };
-    let value_pass_arg = || {
-        Arg::with_name("value_pass")
-            .long("value_pass")
-            .required(false)
-            .takes_value(true)
-            .help("A value password to use")
-    };
-    let verbose_arg = || {
-        Arg::with_name("verbose")
-            .short("v")
-            .multiple(true)
-            .help("Sets the level of verbosity")
+enum Storage {
+    Memory(MemStore<Sha256Digest>),
+    Ipfs(IpfsStore),
+    Sqlite(SqliteStore),
+}
+impl ReadOnlyStore<Sha256Digest> for Storage {
+    fn get(&self, link: &Sha256Digest) -> Result<Box<[u8]>> {
+        match self {
+            Self::Memory(m) => m.get(link),
+            Storage::Ipfs(i) => i.get(link),
+            Storage::Sqlite(s) => s.get(link),
+        }
+    }
+}
+
+impl BlockWriter<Sha256Digest> for Storage {
+    fn put(&self, data: Vec<u8>) -> Result<Sha256Digest> {
+        match self {
+            Self::Memory(m) => m.put(data),
+            Storage::Ipfs(i) => i.put(data),
+            Storage::Sqlite(s) => s.put(data),
+        }
+    }
+}
+impl FromStr for Storage {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self> {
+        let s = match s {
+            "memory" => Self::Memory(MemStore::new(usize::max_value(), Sha256Digest::new)),
+            "ipfs" => Self::Ipfs(IpfsStore::new()?),
+            x => Self::Sqlite(SqliteStore::new(x)?),
+        };
+        Ok(s)
+    }
+}
+#[derive(StructOpt)]
+#[structopt(about = "CLI to work with large banyan trees on ipfs")]
+struct Opts {
+    #[structopt(long, global = true)]
+    /// An index password to use
+    index_pass: Option<String>,
+    #[structopt(long, global = true)]
+    /// A value password to use
+    value_pass: Option<String>,
+    #[structopt(short, parse(from_occurrences = set_log_level), global = true)]
+    #[allow(dead_code)] // log level will bet set in [`set_log_level`]
+    /// Increase verbosity
+    verbosity: u64,
+    #[structopt(long, default_value = "memory", global = true)]
+    /// Storage, possible options "memory", "ipfs" (uses local ipfs gw via HTTP),
+    /// or a path to a sqlite database (will create it if it doesn't exist)
+    storage: Storage,
+    #[structopt(subcommand)]
+    cmd: Command,
+}
+
+fn set_log_level(verbosity: u64) -> u64 {
+    let level = match verbosity {
+        0 => Level::ERROR,
+        1 => Level::INFO,
+        2 => Level::DEBUG,
+        _ => Level::TRACE,
```

Comment:
does this effectively default to 0 or to trace? (`verbosity` doesn't have `default_value`.)

Reply:
0